Automattic/newspack-newsletters
# feat: support send lists in ESP classes

### All Submissions:






### Changes proposed in this Pull Request:



To be reviewed after #1629 (the base for this PR).

- Adds `get_send_lists()` methods to all ESP classes and to the `Newspack_Newsletters_ESP_API_Interface` schema
- Refactors `get_lists()` and `get_segments()` methods for some ESPs to allow for query args (so we only fetch those we need to, where the ESP's API supports filtering by name or IDs)
- Standardizes `senderEmail`, `senderName`, `send_list_id` and `send_sublist_id` meta fields to store that info in a standard format across all ESPs
- Deprecates and migrates provider-specific meta fields that are no longer needed due to the standardized fields
- Refactors `retrieve` and `sync` methods for all ESPs for greater consistency in the `newsletterData` output (all ESPs will now include `lists` and `sublists` keys to store send lists fetched from the ESP)
- Adds caching of Mailchimp audiences in addition to groups/tags/segments

### How to test the changes in this Pull Request:

1. In `wp shell`, try out the new `get_send_lists()` methods. e.g. for MC:

```bash
> $esp = Newspack_Newsletters::get_service_provider_instance( 'mailchimp' )
= Newspack_Newsletters_Mailchimp {#2377
    +service: "mailchimp",
    +name: "Mailchimp",
    +controller: Newspack_Newsletters_Mailchimp_Controller {#2378},
  }

> $esp->get_send_lists()

> $esp->get_send_lists( [ 'type' => 'sublist' ] )

> $esp->get_send_lists( [ 'ids' => [ '123, '456' ] ] ) // where the IDs are actual list or sublist IDs from your connected ESP account
```

2. For MC, trigger the cache refresh and ensure that audience data is cached:

```bash
> wp cron event run newspack_nl_mailchimp_refresh_cache
Executed the cron event 'newspack_nl_mailchimp_refresh_cache' in 1.023s.

> wp option get newspack_nl_mailchimp_cache_lists
array (
  0 =>
  array (
    'id' => '123',
    'web_id' => 456
    'name' => 'My Audience',
    'stats' =>
    array (
      'member_count' => 47,
    ),
  ),
)
```

3. Check out the new `retrieve` output:

```bash
> $esp = Newspack_Newsletters::get_service_provider_instance( 'mailchimp' )
= Newspack_Newsletters_Mailchimp {#2377
    +service: "mailchimp",
    +name: "Mailchimp",
    +controller: Newspack_Newsletters_Mailchimp_Controller {#2378},
  }

> $data = $esp->retrieve( $post_id ) // where $post_id is a newsletter post
```

4. Ensure that the `retrieve` output includes `list` and `sublist` params:

```bash
> $data['lists']
= [
    Newspack\Newsletters\Send_List {#14951
      +"provider": "mailchimp",
      +"type": "list",
      +"entity_type": "audience",
      +"id": "123",
      +"name": "My Audience",
      +"count": 47,
      +"edit_link": "https://us10.admin.mailchimp.com/audience/contacts/?id=456",
      +"value": "123",
      +"label": "[AUDIENCE] My Audience (47 contacts)",
    },
  ]

> $data['sublists']
= [
    Newspack\Newsletters\Send_List {#14911
      +"provider": "mailchimp",
      +"type": "sublist",
      +"entity_type": "group",
      +"id": "789",
      +"name": "My Group",
      +"count": 15,
      +"edit_link": "https://us10.admin.mailchimp.com/audience/groups/?id=654",
      +"value": "789",
      +"label": "[GROUP] My Group (15 contacts)",
    },
  ]
```

### Other information:





---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207717004877591
